CSHARP-5171: Add IReadOnlyDictionary to existing IDictionary indexer translation

damieng opened the pull request:
Fixes CSHARP-5171.

rstam commented on tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp5171Tests.cs:
```diff
@@ -0,0 +1,204 @@
+        [ParameterAttributeData]
+        public void Select_ReadOnlyDictionary_item_with_string_using_compiler_generated_expression_should_work(
+            [Values(LinqProvider.V2, LinqProvider.V3)]
+            LinqProvider linqProvider)
```
Same for other test methods below.

damieng replied:
Yes, IDE did this on paste. Not sure if it's coming from .editorconfig or elsewhere.

rstam commented on tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp5171Tests.cs:
```diff
@@ -0,0 +1,204 @@
+                .Select(x => x.Dictionary["a"]);
+
+            var stages = Translate(collection, queryable);
+            var expectedStage = linqProvider == LinqProvider.V2 ? "{ $project : { a : '$Dictionary.a', _id : 0 } }" : "{ $project : { _v : '$Dictionary.a', _id : 0 } }";
```
Same for other test methods below.

damieng replied:
Same.

rstam commented on tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp5171Tests.cs:
```diff
@@ -0,0 +1,204 @@
+            var collection = GetCollection(linqProvider);
+            var x = Expression.Parameter(typeof(C), "x");
+            var body = Expression.Call(
+                Expression.Property(x, typeof(C).GetProperty("Dictionary")!),
```
The original file didn't have `!` in it.

I understand what it does, but we haven't started using it generally yet and it's not required to compile without warnings.

Same for other cases below.

damieng replied:
So you want to add new code with warnings?

rstam replied:
No I did not say that. I said that it compiles without warnings WITHOUT using `!`.

Are your IDE settings different from ours?

rstam commented on tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp5171Tests.cs:
```diff
@@ -0,0 +1,204 @@
+                typeof(IReadOnlyDictionary<string, int>).GetProperty("Item")!.GetGetMethod(),
+                Expression.Constant("a"));
+            var parameters = new[] {x};
+            var selector = Expression.Lambda<Func<C, int>>(body, parameters);
```
You simplified `new ParameterExpression[]` to `new[]`.

We could simplify even further by using the newish C# syntax for array values:

    var selector = Expression.Lambda<Func<C, int>>(body, [x]);

rstam replied:
Same for several other places if you are interested in making this change.

rstam commented on tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp5171Tests.cs:
```diff
@@ -0,0 +1,204 @@
+            CreateCollection(
+                collection,
+                new C {Id = 1, Dictionary = new ReadOnlyDictionary<string, int>(new Dictionary<string, int> {["a"] = 1})},
+                new C {Id = 2, Dictionary = new ReadOnlyDictionary<string, int>(new Dictionary<string, int> {["b"] = 2})});
```
Your formatting is different than the driver conventions regarding spaces after `{` and before `}`

    CreateCollection(
        collection,
        new C { Id = 1, Dictionary = new ReadOnlyDictionary<string, int>(new Dictionary<string, int> { ["a"] = 1 }) },
        new C { Id = 2, Dictionary = new ReadOnlyDictionary<string, int>(new Dictionary<string, int> { ["b"] = 2 }) });

Did your IDE do this?

damieng replied:
Yes. I've not manually formatted anything manually. Just letting my editor pickup the .editconfig settings.

rstam reviewed (changes requested):
In general looks good and correct.

I've suggested some changes for you to consider.

damieng commented on tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp5171Tests.cs:
```diff
@@ -0,0 +1,204 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
I believe it's picking up from the .editorconfig which seems to be configured wrong to preserve the existing style. e.g. it specifies `s_membername` for static private members but we're actually using `__membername`.

rstam replied:
Use standard indentation on multiline comment.

rstam replied:
Looks like you need to sort out your IDE settings.

And also looks like we need to sort out some `.editorconfig` settings.

I don't know why our IDEs are behaving differently than yours.

damieng commented:
Addressed all the feedback except:

1. The ! to indicate not null - really don't want to check in new code with compiler warnings.
2. Left the method as-is after the slack discussion. Can still change to original style if preferred.

rstam commented on src/MongoDB.Driver/Linq/Linq3Implementation/Misc/TypeExtensions.cs:
```diff
@@ -1,17 +1,17 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Don't change formatting of existing comments.

Looks like you need to figure out your IDE settings.

rstam commented on src/MongoDB.Driver/Linq/Linq3Implementation/Misc/TypeExtensions.cs:
```diff
@@ -255,28 +261,14 @@ public static bool IsValueTuple(this Type type)
-            return false;
+            genericInterface = type.IsConstructedGenericType && interfaceDefinitions.Contains(type.GetGenericTypeDefinition())
+                ? type
+                : type.GetInterfaces().FirstOrDefault(i => i.IsGenericType && interfaceDefinitions.Contains(i.GetGenericTypeDefinition()));
```
Use `IsConstructedGenericType` on this line also.

rstam commented on tests/MongoDB.Driver.Tests/Linq/Linq3Implementation/Jira/CSharp5171Tests.cs:
```diff
@@ -0,0 +1,198 @@
+            var body = Expression.MakeIndex(
+                Expression.Property(x, typeof(C).GetProperty("Dictionary")!),
+                typeof(IReadOnlyDictionary<string, int>).GetProperty("Item"),
+                new Expression[] {Expression.Constant("a")});
```
You switched to `[x]` on the next line. Should we standardize on that notation in all relevant places in this file. For example, this line could be:

    [Expression.Constant("a")]);

rstam commented:
> Addressed all the feedback except:
> 
> 1. The ! to indicate not null in one test - really don't want to check in new code with compiler warnings.

Agreed that we don't want to check in new code with compiler warnings.

However, I was NOT getting any compiler warnings. Perhaps your IDE settings are different than ours?
